### ISSUE TITLE
I am facing an issue on embedded app in iphone safari in login section

### ISSUE THREAD

**Issue by azarrar** [osiset/laravel-shopify#277]
When i connect to myApp using shopify admin section, App is working but when i use login section of laravel it is showing 419 error of laravel only in apple safari. but working find in all browser chorme, firefox.

**Comment by azarrar** [comment]
resolved by enabling cookies in safari.

**Comment by bilfeldt** [comment]
The problem is third party cookies which is actually disabled in Safari per default. Check your browser here: [https://www.whatismybrowser.com/detect/are-third-party-cookies-enabled](https://www.whatismybrowser.com/detect/are-third-party-cookies-enabled)

Actually, this is can be a problem for many users since Safari is pretty widely used and this is a default setting.

@ohmybrew and @azarrar I suggest re-opening this issue. 

The only way to solve this would probably be to rely on sessions instead of third party cookies. I am not sure it that can be done @ohmybrew? But at least we can add a middleware that can detect this so that it's possible to show a more detailed page with information about how to enable this in Safari.

What do you think?

**Comment by ohmybrew** [comment]
To my knowledge there is nothing I can do on my end @bilfeldt :( While the idea of showing a message is nice, I think its something not needed, I don't want to have a bunch of code for every situation like this.

**Comment by bilfeldt** [comment]
@ohmybrew This is caused by the Oauth process. There is a question about this on the Shopify forum [here](https://community.shopify.com/c/Shopify-APIs-SDKs/How-to-enable-third-party-cookies-embedded-apps/td-p/439425).

Basically this 419 error is caused by the app setting the cookie from within the iFrame which is blocked by browsers like Safari as part of ITP 2.0, since the cookie domain does not correspond to the parent domain - see Github discussion [here](https://github.com/Shopify/shopify_app/issues/533#issuecomment-361683084)

Shopify has created an entire page with recommended flow for handling ITP 2.0, which can be seen [here](https://help.shopify.com/en/api/guides/itp-impact). This is the recommended flow:

![image](https://user-images.githubusercontent.com/19682/45030452-da6c4600-b019-11e8-8dcf-68b0aadd17a2.png)

I think it would be a great addition to the plugin, to handle this for Safari (preferably a setting which browsers to do this for) since Safari is a widely used browser and it seems this might be implemented by other browsers as well (I noted that it seems that some users are experiencing this on some versions of Firefox as well, but havn't confirmed if it is the same error).

Note from the Shopify admin [comment](https://community.shopify.com/c/Shopify-APIs-SDKs/How-to-enable-third-party-cookies-embedded-apps/td-p/439425) that the redirect page must be a blank page:

> if the app renders UI during this redirect step, it creates visible flashes of content that can be confusing for the merchant. Instead, we recommend rendering a plain blank/white page during the Oauth redirect steps.

What is your thoughts on this @ohmybrew?

**Comment by ohmybrew** [comment]
I already do a full page redirect when doing the oauth, what exactly is it I need to do you think, @bilfeldt ?

**Comment by bilfeldt** [comment]
@ohmybrew I have tried to read the Wiki section [Shop Installation and Authentication Process](https://github.com/ohmybrew/laravel-shopify/wiki/Shop-Installation-and-Authentication-Process), but I was not completely sure which steps occur within the iFrame (embedded) and which occur in Top-level.

Are you sure that the cookie is being set in top-level? If that is the case then I don't get why we are seeing this 419 error which should be because the cookie is being set from the embedded iFrame.

I also looked at #293 and especially your [comment](https://github.com/ohmybrew/laravel-shopify/issues/271#issuecomment-509922990) on #271 where you describe the flow for the partial authentication. Maybe it would be nice with a diagram? I can easily add a diagram to the Wiki if you would be able to provide detailed information about the flow?

**Comment by bilfeldt** [comment]
@ohmybrew I can also confirm that this issue is also present on the official [Shopify App](https://apps.apple.com/us/app/shopify-ecommerce-business/id371294472) for iPhone and iPad. 

When using the app, then you are facing the same authentication screen and once entering the shopify domain (which is already in the system) then a 419 page is shown.

I am pretty sure this is caused by the same restriction since the Shopify app probably uses the Safari browser under the hood.

I have also tested this on [Shopify's Android app](https://play.google.com/store/apps/details?id=com.shopify.mobile&hl=en_US) where there is no problem - assumably because it's using the Chrome browser under the hood.

**Comment by ohmybrew** [comment]
Great research @bilfeldt :)
Is there an easy work around to it if even the official app isnt handling it well?

Looking at your diagram from before... how would we even reliably test the cookie since its a header which is read by the browser... we would have to have a page which sets a cookie, then go to another page to try and read the cookie? Seems like a huge mess :(

**Comment by bilfeldt** [comment]
@ohmybrew Thanks :) I have tried to debug the `AuthShop` middleware using Safari, please see below.

## Setup
So I have added the following debug lines to `/src/ShopifyApp/Middleware/AuthShop.php`

```php
/**
 * Handle an incoming request.
 *
 * @param \Illuminate\Http\Request $request
 * @param \Closure                 $next
 *
 * @return mixed
 */
public function handle(Request $request, Closure $next)
{
    $validation = $this->validateShop($request);
    if ($validation !== true) {
        Log::debug('Validation not TRUE');
        return $validation;
    }
    Log::debug('Validation TRUE');

    return $next($request);
}
```

```php
/**
 * Checks we have a valid shop.
 *
 * @param \Illuminate\Http\Request $request
 *
 * @return bool|\Illuminate\Http\RedirectResponse
 */
protected function validateShop(Request $request)
{
    // Setup the session service
    $session = new ShopSession();

    // Grab the shop's myshopify domain from query or session
    $shopDomainParam = $request->get('shop');
    $shopDomainSession = $session->getDomain();
    $shopDomain = ShopifyApp::sanitizeShopDomain($shopDomainParam ?? $shopDomainSession);
    Log::debug('Shop domain from request: ' . $shopDomainParam);
    Log::debug('Shop domain from session: ' . $shopDomainParam);
    Log::debug('Shop domain: ' . $shopDomainParam);

    // Get the shop based on domain and update the session service
    $shopModel = Config::get('shopify-app.shop_model');
    $shop = $shopModel::withTrashed()
        ->where(['shopify_domain' => $shopDomain])
        ->first();

    $session->setShop($shop);

    $flowType = null;
    if ($shop === null || $shop->trashed()) {
        // We need to do a full flow
        $flowType = AuthShopHandler::FLOW_FULL;
        Log::debug('Doing flow: ' . $flowType);
    } elseif (!$session->isValid()) {
        // Just a session issue, do a partial flow if we can...
        Log::debug('Cannot decide about flow yet');
        $flowType = $session->isType(ShopSession::GRANT_PERUSER) ?
            AuthShopHandler::FLOW_FULL :
            AuthShopHandler::FLOW_PARTIAL;
        Log::debug('Doing flow: ' . $flowType);
    }

    if ($flowType !== null) {
        // We have a bad session
        Log::warning('BAD FLOW');
        return $this->handleBadSession(
            $flowType,
            $session,
            $request,
            $shopDomain
        );
    }

    // Everything is fine!
    return true;
}
```

and lastly

```php
/**
 * Handles a bad shop session.
 *
 * @param string                                    $type       The auth flow to perform.
 * @param \OhMyBrew\ShopifyApp\Services\ShopSession $session    The session service for the shop.
 * @param \Illuminate\Http\Request                  $request    The incoming request.
 * @param string|null                               $shopDomain The incoming shop domain.
 *
 * @return \Illuminate\Http\RedirectResponse
 */
protected function handleBadSession(
    string $type,
    ShopSession $session,
    Request $request,
    string $shopDomain = null
) {
    Log::debug('Current session domain: ' . $session->getDomain());
    // Clear all session variables (domain, token, user, etc)
    $session->forget();

    // Set the return-to path so we can redirect after successful authentication
    Session::put('return_to', $request->fullUrl());

    // Depending on the type of grant mode, we need to do a full auth or partial
    Log::debug('New session domain: ' . $session->getDomain());
    Log::debug('Flow type: ' . $type);
    return Redirect::route(
        'authenticate',
        array_merge(
            $request->all(),
            [
                'type' => $type,
                'shop' => $shopDomain,
            ]
        )
    );
}
```

## Process
Before this, please clear all browser cookies and cache. Also make sure that the app is already installed once and hence created in the `shops` table.

Process to test:
- Login to Shopify
- Go to "Apps"
- Click on the app

## Result
This is the log
```
[2019-08-09 13:58:13] local.DEBUG: Shop domain from request: foo-bar.myshopify.com  
[2019-08-09 13:58:13] local.DEBUG: Shop domain from session: foo-bar.myshopify.com  
[2019-08-09 13:58:13] local.DEBUG: Shop domain: foo-bar.myshopify.com  
[2019-08-09 13:58:13] local.DEBUG: Cannot decide about flow yet  
[2019-08-09 13:58:13] local.DEBUG: Doing flow: partial  
[2019-08-09 13:58:13] local.WARNING: BAD FLOW  
[2019-08-09 13:58:13] local.DEBUG: Current session domain:   
[2019-08-09 13:58:13] local.DEBUG: New session domain:   
[2019-08-09 13:58:13] local.DEBUG: Flow type: partial  
[2019-08-09 13:58:13] local.DEBUG: Validation not TRUE  
[2019-08-09 13:58:13] local.DEBUG: Shop domain from request:   
[2019-08-09 13:58:13] local.DEBUG: Shop domain from session:   
[2019-08-09 13:58:13] local.DEBUG: Shop domain:   
[2019-08-09 13:58:13] local.DEBUG: Doing flow: full  
[2019-08-09 13:58:13] local.WARNING: BAD FLOW  
[2019-08-09 13:58:13] local.DEBUG: Current session domain:   
[2019-08-09 13:58:13] local.DEBUG: New session domain:   
[2019-08-09 13:58:13] local.DEBUG: Flow type: full  
[2019-08-09 13:58:13] local.DEBUG: Validation not TRUE  
```

Now I see the screen "Enter your Shopify domain to login" **inside** the iFrame with an empty input field, which in my opinion makes no sense since this shop exist? Also, maybe we should prefill that field if the domain matches the format `'%'.config('shopify-app.myshopify_domain'')`  - separate improvement?

- Fill out the domain (without protocol, so `foo-bar.myshopify.com`) and clicking "submit"

Causes the 419 error without anything else being written to the log.

**Comment by bilfeldt** [comment]
@ohmybrew would you be okay with reopening this issue, so that other might assist with this issue?

**Comment by ohmybrew** [comment]
@bilfeldt Done, yeah so its redirecting to update the session and fails to do so due to Safaris security limitations it seems, thus the shop is empty everywhere.

**Comment by bilfeldt** [comment]
@ohmybrew during my test, the app never escaped the embedded frame. Since it never escaped from the embedded frame and went in to top level, then Safari blocks the session.

How would you redirect to a route in top-level from your auth middleware? It would be nice to test the proposed solution from Shopify, which required that we set a demo cookie first and then check if we can read it. If we are not able to read it, then we break out from embedded to top level where we set a cookie and the do it all again.

**Comment by ohmybrew** [comment]
Hmm yeah it only does a full page redirect on a initial auth... not sure what to do here.

**Comment by bilfeldt** [comment]
@ohmybrew Do you by any chance have a diagram of the auth flow? Or can you confirm that the latest 9.2 release is still using the flow described [here](https://github.com/ohmybrew/laravel-shopify/wiki/Shop-Installation-and-Authentication-Process)? That would really help in case we need to update the flow to account for ITP 2.x

**Comment by ohmybrew** [comment]
Hey!Currently in master, which should represent the next release (possible tonight), the partial flow stuff is removed. So the regular flow just appliea right now. I can draw a diagram when I'm at computer On Aug. 29, 2019 19:27, Anders Bilfeldt <notifications@github.com> wrote:@ohmybrew Do you by any chance have a diagram of the auth flow? Or can you confirm that the latest 9.2 release is still using the flow described here? That would really help in case we need to update the flow to account for ITP 2.x

**Comment by ohmybrew** [comment]
It seems Shopify is updating their internal apps to support this.

I noticed when loading Scripts Editor (Plus app) the other day, it had a embedded page letting me know it had to do this to allow for cookies...

I am not sure the best way to implement such a thing atm for the package.. suggestions welcome.

**Comment by jedimdan** [comment]
The app that I was working on was having session issues on v9 when I attempted to launch it on Safari, but it seems that the issues have gone away on v10 and the app is launching as normal.

**Comment by darrynten** [comment]
This problem exists in Safari on Mac too.

I came across this while investigating a ticket from one of our users who was unable to load the app.

**Comment by bilfeldt** [comment]
It [seem that Shopify is not taking any official steps to fix this from their site](https://community.shopify.com/c/Shopify-APIs-SDKs/ITP-Embedded-App-Update/m-p/571813/highlight/true#M38177). The official Shopify guide is to implement the test cookie flow as explained in their [IPT guide](https://help.shopify.com/en/api/guides/itp-impact).

This should be something that we can implement in the `AuthShop` middleware but will require a blade file rendering which uses EASDK to redirect to a top-level site (outside iFrame) as mentioned:

> Fortunately, the Embedded App SDK (EASDK) will allow you to use the remoteRedirect method to change the location of the parent window. This allows you to redirect the merchant to your app as a parent window, set a cookie, and then redirect back to Shopify to set the cookie in the embedded context.

I will happily make a proposed flow for this. But @ohmybrew can you maybe provide a diagram or at least a bullet list of the current `AuthShop` flow? I can see that you are also working on implementing changes to this here #327 . I am especially interested in the difference between the flows `Offline Grant Mode` and `Per-User Grant Mode` and the motivation for having two flows - maybe this can be simplified?

I look much forward to hearing from you.

**Comment by ncpope** [comment]
I don't think it should be within the AuthShop middleware since some apps are not embedded and those non-embedded apps don't have this issue.

**Comment by bilfeldt** [comment]
@ncpope that makes sense.

I think it would be okay to assume any application having AppBridge enabled is using embedded app and any application not using AppBridge is not embedded. Would you agree?

```php
config('shopify-app.appbridge_enabled'); //FALSE: hosting the app inside ones own container.
```
It should be noted that [Shopify Admin Links](https://help.shopify.com/en/api/embedded-apps/app-extensions/shopify-admin/app-admin-access) will always be applied within the iFrame however.

**Comment by ncpope** [comment]
@bilfeldt 

> config('shopify-app.appbridge_enabled'); 

I think it is safe to say that anyone using this setting is going to be relying on the package to generate their embedded app views. I think this would make it a good setting to compare against when deciding whether or not to implement the fix you found from Shopify.

> It should be noted that Shopify Admin Links will always be applied within the iFrame however.

I am not entirely sure what you mean here. If you are saying that admin links need to be directed to an embedded app, that isn't the case currently. I am using admin links to load links to non-embedded apps/non-iframe apps.

I think you might be referring to: https://help.shopify.com/en/api/embedded-apps/app-extensions/navigation/create-navigation-link

If that's the case then yep you are correct. You can only use those Navigation Links within embedded apps

If you are sure you mean Admin Links, then you just need to double check that your embedded app setting is turned off:

<img width="1027" alt="Screen Shot 2019-09-22 at 12 43 10" src="https://user-images.githubusercontent.com/1903724/65386018-90462780-dd36-11e9-8a00-298a02be4003.png">

**Comment by ohmybrew** [comment]
Shopify's guide seems like there is a lot to implement.. it seems like it would introduce issues.

**Comment by ohmybrew** [comment]
Read over it again.. seems like a lot of extra steps to plug in, is it actually this involved or are they just being very verbose in the docs?

**Comment by andreuka** [comment]
> Read over it again.. seems like a lot of extra steps to plug in, is it actually this involved or are they just being very verbose in the docs?

Hi there, @ohmybrew !

As I can understand from there and apps which already improved this, we just need a page which will have JavaScript included, which will check if the safari browser already allowed to store 3rd party cookies, if it not allows request access to 3rd party cookies and after user approve check again and if we have right to store cookies already, we just need immediate redirect to next page.

So if browser is safari, we need to send them always to this "safari auth page", before they can access any other app logic.

Does it makes sense?

Thanks!

**Comment by ohmybrew** [comment]
That seems like we would *always* have to show this confirmation page since we need to use JS to verify the cookie status?


```
function cookieStatus() {
    if (navigator.cookieEnabled) {
       return true;
    }

    const cookieKey = 'laravel_shopify';
    document.cookie = `${cookieKey}=1`;

    const staus = document.cookie.indexOf(`${cookieKey}=`) != -1;
    document.cookie = `${cookieKey}=1; expires=Thu, 01-Jan-1970 00:00:01 GMT`;

    return status;
}
```

If cookieStatus is bad => show a page like this at top-level: https://imgur.com/a/GkTeJoH (clicking button sets the testcookie?

If cookieStatus is good => redirect to asked page like normal

Does that sound right?

**Comment by andreuka** [comment]
Hi there again!

I believe this is almost exactly what needed.
I know only one app which already working with new ITP using Safari, but it works in abit strange way - always asking for permissions.

This is the app: https://apps.shopify.com/facebook-marketing

I believe we can do all code we need just in 1 file:
/ohmybrew/laravel-shopify/src/ShopifyApp/resources/views/auth/fullpage_redirect.blade.php

![image](https://user-images.githubusercontent.com/6764188/65536743-ae658080-df0c-11e9-982e-6c6584fc406f.png)

I didnt tested if this will 100% work yet.

**Comment by ohmybrew** [comment]
Thats possible yup. The only app I work with daily that gives me the screen is Script Editor (for plus).

Is there a browser besides Safari that does this... I'm on Linux and Android... no Safari to test this stuff out.

**Comment by andreuka** [comment]
> Thats possible yup. The only app I work with daily that gives me the screen is Script Editor (for plus).

The only browser currently introduced itp 2.1 is safari as I can see.

I can try to setup virtual machine and give you teamviewer if that helps.

**Comment by andreuka** [comment]
Also found this script:
https://cdn.shopifycloud.com/facebook_ads/assets/shopify_app/request_storage_access-3c493a4239442bfd86e4df1c8f6b59d1a71ab911fbc55a8728bbe0cebf494949.js

shopify using this to detect if browser is affected, to request access, etc.

Im trying to integrate this as well.

**Comment by ohmybrew** [comment]
Great, actually I got lambdatest account, I can use that for safari

**Comment by bilfeldt** [comment]
What about checking [test cookie existence](https://laravel.com/docs/master/requests#cookies) in a `testCookieMiddleware` using
```php
Cookie::has('testCookie');
```
If the `testCookie` is not set, then render a view `top-level-redirect.blade.php` which uses [Shopify App Bridge redirect](https://help.shopify.com/en/api/embedded-apps/app-bridge/actions/navigation/redirect#redirect-to-an-absolute-url-outside-of-the-app-and-outside-of-shopify-admin)

```js
redirect.dispatch(Redirect.Action.REMOTE, 'https://my-app.example.com/session/set-top-level-cookie');
```
to redirect the user to a new endpoint `/session/set-top-level-cookie` in **top-level** (not embedded). This endpoint will be able to set the `testCookie` without any issues since this page is not loaded in an embedded frame:
```php
/**
 * Set the testCookie and redirect to home-redirect
 *
 * @return Response
 */
public function index()
{
    Cookie::queue('testCookie', true, 15);

    return view('home-redirect');
}
```
The `home-redirect.blade.php` will simply show something like
> App is now authenticated (IPT requirements). You will be redirected shortly and should not see this error again.

The blade then redirects the user to the home route and everything will start over again except that the `testCookieMiddleware` will not pass since the cookie is set.

**Comment by bilfeldt** [comment]
Just for information then we have also gotten reports from some customers using the latest Firefox browser that our app is blocked. It seems that Firefox is also [blocking 3rd party cookies now](https://blog.mozilla.org/blog/2019/06/04/firefox-now-available-with-enhanced-tracking-protection-by-default/#targetText=One%20of%20those%20initiatives%20outlined,to%20track%20their%20every%20move.). We have not yet confirmed this but this is a setting in Chrome can we see.

**Comment by ohmybrew** [comment]
I have created a branch [itp-feature]. I'll begin work on it and check it out...

**Comment by ohmybrew** [comment]
Having some success thus far as a quick update... https://i.imgur.com/soM3bj3.png

**Comment by ohmybrew** [comment]
Have base code working... clicking continue sets the test cookie (configurable name) and redirects to home where authshop allows it through... going to boot up a Safari instance to test now.

**Comment by jedimdan** [comment]
I'm on the Shopify Partners slack, and some partners have reported that this is starting to happen on Chrome too. Partners report multiple embedded shopify apps going into an auth loop, including even Shopify's own GraphQL testbed app.

**Comment by ptcampbell** [comment]
I've experienced this too, including with Firefox. Looks like an experimental flag in Chrome causing the issue right now (#356)

**Comment by jedimdan** [comment]
@ptcampbell the concern if this experimental flag will/has become non-experimental and part of a stable Chrome build

**Comment by jedimdan** [comment]
Just an update regarding the Chrome auth loop problem (does this need to be a different issue?): 

A rep from Shopify shared on the partners slack that cookies from embedded apps need to be created with `SameSite=None; Secure`. [Here's the reference on the Chromium blog](https://blog.chromium.org/2019/10/developers-get-ready-for-new.html).

We went ahead and changed our app's `config/session.php` to update `'secure' => true,` and `'same_site' => 'none',` and have confirmed that the auth loop no longer occurs on Chrome 80.

Now the question is should this be in the installation documentation, or should this package go ahead and force this setting in the service provider?

**Comment by jedimdan** [comment]
I went ahead and created a new issue #367

**Comment by darrynten** [comment]
This is going to become an issue in 5 days time and will affect all of our apps.

I'm going to jump on resolving the `SameSite=None; Secure` requirement.

**Comment by Kyon147** [comment]
This and the bee parameter change need to be pushed to master as proper fixes as a lot of people will have their app breaks.

**Comment by mferrario** [comment]
> This is going to become an issue in 5 days time and will affect all of our apps.
> 
> I'm going to jump on resolving the `SameSite=None; Secure` requirement.

What about the solution that @jedimdan proposed?

**Comment by darrynten** [comment]
@mferrario 

I've made the changes that @jedimdan suggested and can confirm they work.

I have also investigated it a bit deeper, and it appears that older browsers will not be compatible with this change.

This is why there needs to be a change somewhere/somehow that won't set this for older browsers.

There is an open PR proposed by @jedimdan - #375 

I'm currently going through this and will propose some changes.

Once this issue is resolved I'll jump onto looking at the new signature verification failure problem too.

**Comment by jedimdan** [comment]
@darrynten not sure if you're caught up on all the conversations regarding the signature verification problem, but it's already solved in the `master` branch (Refer to #359). The problem is that those changes are not released yet. Just FYI so that no double work is done.

**Comment by darrynten** [comment]
See #382 for a fix

**Comment by darrynten** [comment]
@jedimdan thanks for the heads up! Have pulled latest master and fix is indeed in place.
